Install MDB entry in all vlans when vlan is not specified and vlan filtering
is enabled.

diff --git a/net/bridge/br_mdb.c b/net/bridge/br_mdb.c
index 80f648f..c0b8bbc 100644
--- a/net/bridge/br_mdb.c
+++ b/net/bridge/br_mdb.c
@@ -482,6 +482,10 @@ static int br_mdb_add(struct sk_buff *skb, struct nlmsghdr *nlh, void *arg)
 	struct net_device *dev;
 	struct net_bridge *br;
 	int err;
+	struct net_port_vlans *pv;
+	unsigned short vid = VLAN_N_VID;
+	struct net_device *pdev;
+	struct net_bridge_port *p;
 
 	err = br_mdb_parse(skb, nlh, &dev, &entry);
 	if (err < 0)
@@ -489,9 +493,30 @@ static int br_mdb_add(struct sk_buff *skb, struct nlmsghdr *nlh, void *arg)
 
 	br = netdev_priv(dev);
 
-	err = __br_mdb_add(net, br, entry);
-	if (!err)
-		__br_mdb_notify(dev, entry, RTM_NEWMDB);
+	/* If vlan filtering is enabled and VLAN is not specified
+	 * install mdb entry on all vlans configured on the port.
+	 */
+	pdev = __dev_get_by_index(net, entry->ifindex);
+	if (!pdev)
+		return -ENODEV;
+
+	p = br_port_get_rtnl(pdev);
+	if (!p || p->br != br || p->state == BR_STATE_DISABLED)
+		return -EINVAL;
+
+	pv = nbp_get_vlan_info(p);
+	if (br->vlan_enabled && pv && (entry->addr.vlan_id == 0)) {
+		for_each_set_bit(vid, pv->vlan_bitmap, VLAN_N_VID) {
+			entry->addr.vlan_id = vid;
+			err = __br_mdb_add(net, br, entry);
+			if (!err)
+				__br_mdb_notify(dev, entry, RTM_NEWMDB);
+		}
+	} else {
+		err = __br_mdb_add(net, br, entry);
+		if (!err)
+			__br_mdb_notify(dev, entry, RTM_NEWMDB);
+	}
 	return err;
 }
 
@@ -562,16 +587,41 @@ static int br_mdb_del(struct sk_buff *skb, struct nlmsghdr *nlh, void *arg)
 	struct br_mdb_entry *entry;
 	struct net_bridge *br;
 	int err;
-
+	struct net *net = sock_net(skb->sk);
+	struct net_port_vlans *pv;
+	unsigned short vid = VLAN_N_VID;
+	struct net_device *pdev;
+	struct net_bridge_port *p;
 	err = br_mdb_parse(skb, nlh, &dev, &entry);
 	if (err < 0)
 		return err;
 
 	br = netdev_priv(dev);
 
-	err = __br_mdb_del(br, entry);
-	if (!err)
-		__br_mdb_notify(dev, entry, RTM_DELMDB);
+	/* If vlan filtering is enabled and VLAN is not specified
+	 * delete mdb entry on all vlans configured on the port.
+	 */
+	pdev = __dev_get_by_index(net, entry->ifindex);
+	if (!pdev)
+		return -ENODEV;
+
+	p = br_port_get_rtnl(pdev);
+	if (!p || p->br != br)
+		return -EINVAL;
+
+	pv = nbp_get_vlan_info(p);
+	if (br->vlan_enabled && pv && (entry->addr.vlan_id == 0)) {
+		for_each_set_bit(vid, pv->vlan_bitmap, VLAN_N_VID) {
+			entry->addr.vlan_id = vid;
+			err = __br_mdb_del(br, entry);
+			if (!err)
+				__br_mdb_notify(dev, entry, RTM_DELMDB);
+		}
+	} else {
+		err = __br_mdb_del(br, entry);
+		if (!err)
+			__br_mdb_notify(dev, entry, RTM_DELMDB);
+	}
 	return err;
 }
 
